src/keys/slip10.rs

```diff
@@ -44,100 +99,404 @@ impl Seed {
+                curve: self.curve(),
+                ext: [0_u8; 65],
+            };
+            k.ext[..33].copy_from_slice(&self.0.calc_public_bytes());
+            k.ext[33..].copy_from_slice(self.chain_code());
+            Ok(ExtendedPublicKey(k))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "curve supporting non-hardened key derivation",
+            })
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ExtendedPublicKey(KeyImpl);
+
+impl ExtendedPublicKey {
+    pub fn try_from_extended_bytes(curve: Curve, ext_bytes: &[u8; 65]) -> crate::Result<Self> {
+        if !curve.is_non_hardened_supported() {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "curve supporting non-hardened key derivation",
+            })
+        } else {
+            let key = KeyImpl { curve, ext: *ext_bytes };
+            if !key.is_public_key_valid() {
+                Err(crate::Error::InvalidArgumentError {
+                    alg: "SLIP10",
+                    expected: "valid extended public key",
+                })
+            } else {
+                Ok(Self(key))
+            }
+        }
+    }
+
+    pub fn curve(&self) -> Curve {
+        self.0.curve
+    }
+
+    pub fn extended_bytes(&self) -> &[u8; 65] {
+        self.0.extended_public_bytes()
+    }
+
+    pub fn public_bytes(&self) -> &[u8; 33] {
+        self.0.public_bytes()
+    }
 
-        Ok(Self(i))
+    pub fn public_key(&self) -> PublicKey {
+        match self.curve() {
+            #[cfg(feature = "ed25519")]
+            Curve::Ed25519 => {
+                unreachable!("SLIP10 does not support non-hardened key derivation from Ed25519 public keys")
+            }
+            #[cfg(feature = "secp256k1")]
+            Curve::Secp256k1 => secp256k1_ecdsa::PublicKey::try_from_bytes(self.public_bytes())
+                .map(PublicKey::Secp256k1Ecdsa)
+                .expect("valid extended public key"),
+        }
+    }
+
+    pub fn chain_code(&self) -> &[u8; 32] {
+        self.0.chain_code()
+    }
+
+    pub fn child_key(&self, segment: &Segment) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        debug_assert_ne!(Curve::Ed25519, self.curve());
+
+        if !segment.is_hardened() {
+            Ok(Self(self.0.child_key(segment)))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "non hardened key index for master public key",
+            })
+        }
     }
 
-    pub fn derive(&self, chain: &Chain) -> crate::Result<Key> {
-        let mut k = *self;
-        for c in &chain.0 {
-            k = k.child_key(c)?;
+    pub fn derive(&self, chain: &Chain) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        debug_assert_ne!(Curve::Ed25519, self.curve());
+
+        if chain.all_non_hardened() {
```

Comment:
Nothing at all has to be hardened for Secp256k1?

Reply:
Child public key can be derived from extended master public key only with non-hardened segment. Hardened segments require knowledge of the master private key which is obviously not available.